File: Java/25_Dynamic Programming/H-TUF-Matrix chain multiplication.java
// TUF: Matrix chain multiplication
// https://takeuforward.org/plus/dsa/problems/matrix-chain-multiplication

/**
 * Given a chain of matrices A1, A2, A3,.....An, you have to figure out the most efficient way to multiply these matrices. 
 * In other words, determine where to place parentheses to minimize the number of multiplications.

 * Given an array nums of size n. Dimension of matrix Ai ( 0 < i < n ) is nums[i - 1] x nums[i].
 * Find a minimum number of multiplications needed to multiply the chain.

 * Examples:
 * Input : nums = [10, 15, 20, 25]
 * Output : 8000
 * Explanation : There are two ways to multiply the chain - A1*(A2*A3) or (A1*A2)*A3.
 * If we multiply in order- A1*(A2*A3), then number of multiplications required are 11250.
 * If we multiply in order- (A1*A2)*A3, then number of multiplications required are 8000.
 * Thus minimum number of multiplications required is 8000.
 * 
 * Input : nums = [4, 2, 3]
 * Output : 24
 * Explanation : There is only one way to multiply the chain - A1*A2.
 * Thus minimum number of multiplications required is 24.
 */

class Solution {
    public int matrixMultiplication(int[] nums) {
        int n = nums.length;
        if (n <= 2) return 0; // No multiplication needed for 0 or 1 matrix
        
        // dp[i][j] will store the minimum number of multiplications needed to multiply 
        // the chain of matrices from i to j (0-based indexing)
        int[][] dp = new int[n][n];
        
        // Fill the dp table in bottom-up manner
        // len is the length of the chain
        for (int len = 2; len < n; len++) {
            // i is the start index of the current chain
            for (int i = 1; i < n - len + 1; i++) {
                int j = i + len - 1; // end index of the current chain
                dp[i][j] = Integer.MAX_VALUE;
                
                // Try all possible partition points between i and j
                for (int k = i; k < j; k++) {
                    // Cost of multiplying (i..k) and (k+1..j) plus cost of multiplying the two resulting matrices
                    int cost = dp[i][k] + dp[k + 1][j] + nums[i - 1] * nums[k] * nums[j];
                    if (cost < dp[i][j]) {
                        dp[i][j] = cost;
                    }
                }
            }
        }
        
        // The result is stored in dp[1][n-1]
        return dp[1][n - 1];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        int[] nums1 = {10, 15, 20, 25};
        System.out.println("Test Case 1: " + solution.matrixMultiplication(nums1)); // Expected: 8000
        
        // Test case 2
        int[] nums2 = {4, 2, 3};
        System.out.println("Test Case 2: " + solution.matrixMultiplication(nums2)); // Expected: 24
        
        // Test case 3: Single matrix (no multiplication needed)
        int[] nums3 = {1, 2};
        System.out.println("Test Case 3: " + solution.matrixMultiplication(nums3)); // Expected: 0
        
        // Test case 4
        int[] nums4 = {1, 2, 3, 4, 5};
        System.out.println("Test Case 4: " + solution.matrixMultiplication(nums4)); // Expected: 38
    }
}

File: JavaScript/25_Dynamic Programming/H-TUF-Matrix chain multiplication.js
/**
 * TUF: Matrix Chain Multiplication
 * https://takeuforward.org/plus/dsa/problems/matrix-chain-multiplication
 *
 * Given dimensions of matrices, find minimum number of multiplications
 * needed to multiply the chain.
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(N³) – Three nested loops
 * Space Complexity | O(N²) – DP table
 */

/**
 * Matrix Chain Multiplication using DP
 * @param {number[]} arr - Array of dimensions
 * @param {number} n - Length of array
 * @returns {number} - Minimum number of multiplications
 */
function matrixMultiplication(arr, n) {
    // dp[i][j] = minimum cost to multiply matrices from i to j
    const dp = Array.from({ length: n }, () => new Array(n).fill(0));
    
    // l is chain length
    for (let l = 2; l < n; l++) {
        for (let i = 1; i < n - l + 1; i++) {
            const j = i + l - 1;
            dp[i][j] = Infinity;
            
            // Try all possible positions to split
            for (let k = i; k < j; k++) {
                const cost = dp[i][k] + dp[k + 1][j] + arr[i - 1] * arr[k] * arr[j];
                dp[i][j] = Math.min(dp[i][j], cost);
            }
        }
    }
    
    return dp[1][n - 1];
}

/**
 * MCM using Recursion with Memoization
 */
function matrixMultiplicationMemo(arr, n) {
    const memo = Array.from({ length: n }, () => new Array(n).fill(-1));
    
    function solve(i, j) {
        if (i >= j) return 0;
        
        if (memo[i][j] !== -1) return memo[i][j];
        
        let minCost = Infinity;
        
        for (let k = i; k < j; k++) {
            const cost = solve(i, k) + solve(k + 1, j) + arr[i - 1] * arr[k] * arr[j];
            minCost = Math.min(minCost, cost);
        }
        
        memo[i][j] = minCost;
        return minCost;
    }
    
    return solve(1, n - 1);
}

/**
 * MCM with parenthesization (show optimal order)
 */
function matrixMultiplicationWithOrder(arr, n) {
    const dp = Array.from({ length: n }, () => new Array(n).fill(0));
    const bracket = Array.from({ length: n }, () => new Array(n).fill(0));
    
    for (let l = 2; l < n; l++) {
        for (let i = 1; i < n - l + 1; i++) {
            const j = i + l - 1;
            dp[i][j] = Infinity;
            
            for (let k = i; k < j; k++) {
                const cost = dp[i][k] + dp[k + 1][j] + arr[i - 1] * arr[k] * arr[j];
                if (cost < dp[i][j]) {
                    dp[i][j] = cost;
                    bracket[i][j] = k;
                }
            }
        }
    }
    
    function printParenthesis(i, j) {
        if (i === j) {
            return `M${i}`;
        }
        return `(${printParenthesis(i, bracket[i][j])} x ${printParenthesis(bracket[i][j] + 1, j)})`;
    }
    
    return {
        minCost: dp[1][n - 1],
        order: printParenthesis(1, n - 1)
    };
}

// Test cases
// Matrices: A1 (10x30), A2 (30x5), A3 (5x60)
const arr = [10, 30, 5, 60];
const n = arr.length;

console.log("Minimum multiplications:", matrixMultiplication(arr, n));
// Expected: 4500
// (A1 x A2) x A3 = 10*30*5 + 10*5*60 = 1500 + 3000 = 4500

console.log("Using Memoization:", matrixMultiplicationMemo(arr, n));

const result = matrixMultiplicationWithOrder(arr, n);
console.log("Min cost:", result.minCost);
console.log("Optimal order:", result.order);

// Another test case
const arr2 = [40, 20, 30, 10, 30];
console.log("\nTest 2:", matrixMultiplication(arr2, arr2.length)); // Expected: 26000